src/synth/fluid_synth_mono.c

```diff
@@ -0,0 +1,683 @@
+ * (PTC).
+ *
+ * @param synth instance.
+ * @param chan MIDI channel number (0 to MIDI channel count - 1).
+ * @param key MIDI note number (0-127).
+ * @param vel MIDI velocity (0-127).
+ * @return FLUID_OK on success, FLUID_FAILED otherwise.
+ */
+int
+fluid_synth_noteon_mono_staccato(fluid_synth_t* synth,int chan,int key,int vel)
+{
+	fluid_channel_t* channel = synth->channel[chan];
+	
+	/* Before playing a new note, if a previous monophonic note is currently
+	   sustained it needs to be released */
+	fluid_synth_release_voice_on_same_note_LOCAL(synth,chan, channel->key_mono_sustained);
+	/* Get possible 'fromkey portamento'   */
+	fluid_synth_get_fromkey_portamento_legato( channel, INVALID_NOTE);
+	/* The note needs to be played by voices allocation  */
+	return fluid_preset_noteon(channel->preset, synth, chan, key, vel);
+}
+
+/**
+ * Plays noteoff for a polyphonic or monophonic note
+ * Please see the description above about "monophonic playing".
+ *
+ *                                       
+ *  noteOff poly >---------------------------------+
+ *                                                 |
+ *                                                 |
+ *                                                 |
+ *  noteoff_mono                             _____\|/_______
+ *  LOCAL        >------------------------->|   noteoff     | 
+ *                                          |   monopoly    |----> noteoff
+ *  Sust.on/off  >------------------------->|_______________|
+ *  Sost.on/off										        
+ *
+ * The function has the same behaviour when the noteoff is poly of mono, except
+ * that for mono noteoff, if any pedal (sustain or sostenuto ) is depressed, the
+ * key is memorized. This is neccessary when the next mono note will be played
+ * staccato, as any current mono note currently sustained will need to be released 
+ * (see fluid_synth_noteon_mono_staccato()).
+ * Note also that for a monophonic legato passage, the function is called only when
+ * the last noteoff of the passage occurs. That means that if sustain or sostenuto
+ * is depressed, only the last note of a legato passage will be sustained.
+ *
+ * @param synth instance.
+ * @param chan MIDI channel number (0 to MIDI channel count - 1).
+ * @param key MIDI note number (0-127).
+ * @param Mono, 1 noteoff on monophonic note.
+ *              0 noteoff on polyphonic note.
+ * @return FLUID_OK on success, FLUID_FAILED otherwise.
+ *
+ * Note: On return, on monophonic, possible sustained note is memorized in
+ * key_mono_sustained. Memorization is done here on noteOff.
+ */
+int fluid_synth_noteoff_monopoly(fluid_synth_t* synth, int chan, int key,
+                            char Mono)
+{
+	int status = FLUID_FAILED;
+	fluid_voice_t* voice;
+	int i;
+	fluid_channel_t* channel = synth->channel[chan];
+	/* Key_sustained is prepared to return no note sustained (-1) */
+	if (Mono)
+	{
+		channel->key_mono_sustained = -1; /* no mono note sustained */
```

Comment:
Why does `key_mono_sustained` use -1 as invalid not and not `INVALID_NOTE`?

Reply:
This is inconsistent , key_mono_sustained should use INVALID_NOTE. This will be corrected.